### PR TITLE
fix: set OUT_DIR for all units with build scripts

### PR DESCRIPTION
**Pull request by hi-rustin**
### What does this PR try to resolve?

close https://github.com/rust-lang/cargo/issues/13127

- Added a minimal case to cover this behavior.
- Set OUT_DIR for all units with build scripts.

### How should we test and review this PR?

Check out the unit test.

### Additional information
None

### CODE REVIEW

File: tests/testsuite/test.rs [hunk 1]
```diff
@@ -4897,3 +4897,51 @@ fn cargo_test_print_env_verbose() {
...
+                        assert!(std::env::var("OUT_DIR").is_ok());
+                        assert_eq!(foo::add(2, 5), 7);
+                    }
+                }
+            "#,
+        )
+        .build();
+
+    p.cargo("test").run();
+    p.cargo("test --package foo --test case -- tests::test_add --exact --nocapture")
```
Comment (hi-rustin) [inline tests/testsuite/test.rs:4943]: It seems that this test is affected by this bug itself. When run independently, it passes. However, if run with 'cargo test -p cargo', it fails.

Comment (hi-rustin) [reply tests/testsuite/test.rs:4943]: Fixed by [`509c75e` (#13204)](https://github.com/rust-lang/cargo/pull/13204/commits/509c75e93126b14cf5cbe28a84afa60778377bef)

**Review by hi-rustin — COMMENTED** [review]
🔢 Self-check

File: tests/testsuite/test.rs [hunk 2]
```diff
@@ -4941,6 +4941,7 @@ fn cargo_test_set_out_dir_env_var() {
 
     p.cargo("test").run();
     p.cargo("test --package foo --test case -- tests::test_add --exact --nocapture")
+        .env_remove("OUT_DIR")
```
Comment (epage) [inline tests/testsuite/test.rs:4944]: Should we do this instead in https://github.com/rust-lang/cargo/blob/master/crates/cargo-test-support/src/lib.rs#L1249 ?

File: tests/testsuite/test.rs [hunk 3]
```diff
@@ -4897,3 +4897,51 @@ fn cargo_test_print_env_verbose() {
         )
         .run();
 }
+
+#[cargo_test]
```
Comment (epage) [inline tests/testsuite/test.rs:4901]: Can you modify the commits so this commit passes tests?

File: src/cargo/core/compiler/context/mod.rs [hunk 4]
```diff
@@ -185,6 +186,32 @@ impl<'a, 'cfg> Context<'a, 'cfg> {
...
+                        .to_string();
+                    let script_meta = self.get_run_build_script_metadata(&dep.unit);
+                    self.compilation
+                        .extra_env
+                        .entry(script_meta)
+                        .or_insert_with(Vec::new)
+                        .push(("OUT_DIR".to_string(), out_dir));
+                }
+            }
+        }
```
Comment (epage) [inline src/cargo/core/compiler/context/mod.rs:213]: If I'm understanding this, we only really need to set `OUT_DIR` once per package and the previous code did that be assuming the lib unit was sufficient (but some packages don't have that).

I'm assuming there isn't a more direct way to iterate over packages to find out the necessary information and set it, and instead we have to walk all units and de-dupe down to individual packages?

File: tests/testsuite/test.rs (hunk 2 above)
Comment (hi-rustin) [reply tests/testsuite/test.rs]: Updated. Good suggestion. ❤️

File: tests/testsuite/test.rs (hunk 3 above)
Comment (hi-rustin) [reply tests/testsuite/test.rs:4901]: Updated.

File: src/cargo/core/compiler/context/mod.rs (hunk 4 above)
Comment (hi-rustin) [reply src/cargo/core/compiler/context/mod.rs:213]: Yes. I think your understanding is correct. Before https://github.com/rust-lang/cargo/pull/3310/files, we set it for every unit. 
image

**Comment by ehuss** [comment]
Thanks!

@bors r+

**Pull request merged**